### PR TITLE
Added rules to encrypt, hash and decrypt fields

### PR DESCRIPTION
**Pull request by KavishShah09**

### CODE REVIEW

File: gateway/modules/auth/match.go [hunk 1]
```diff
@@ -165,3 +176,85 @@ func matchRemove(rule *config.Rule, args map[string]interface{}) (*PostProcess,
...
+				logrus.Errorln("error encrypting value in matchEncrypt: ", err)
+				return nil, err
+			}
+			er := utils.StoreValue(field, encrypted, args)
+			if er != nil {
+				logrus.Errorln("error storing value in matchEncrypt: ", err)
+				return nil, er
+			}
+		} else {
+			return nil, ErrIncorrectRuleFieldType
```
Comment (KavishShah09) [inline gateway/modules/auth/match.go]:
```suggestion
			return nil, fmt.Errorf("invalid field (%s) provided", field)
```

File: gateway/modules/auth/match.go [hunk 2]
```diff
@@ -165,3 +176,85 @@ func matchRemove(rule *config.Rule, args map[string]interface{}) (*PostProcess,
 	}
 	return actions, nil
 }
+
+func (m *Module) matchEncrypt(rule *config.Rule, args map[string]interface{}) (*PostProcess, error) {
+	actions := &PostProcess{}
+	for _, field := range rule.Fields {
+		loadedValue, err := utils.LoadValue(field, args)
```
Comment (KavishShah09) [inline gateway/modules/auth/match.go]: Load value inside else if. Also check if loaded value is string

File: gateway/modules/auth/match.go [hunk 3]
```diff
@@ -165,3 +176,85 @@ func matchRemove(rule *config.Rule, args map[string]interface{}) (*PostProcess,
...
+		if err != nil {
+			logrus.Errorln("error loading value in matchEncrypt: ", err)
+			return nil, err
+		}
+		key, err := base64.StdEncoding.DecodeString(m.secret)
```
Comment (KavishShah09) [inline gateway/modules/auth/match.go]: use a pre decoded m.aeskey here

File: gateway/modules/auth/match.go [hunk 4]
```diff
@@ -165,3 +176,85 @@ func matchRemove(rule *config.Rule, args map[string]interface{}) (*PostProcess,
...
+		if strings.HasPrefix(field, "res") {
+			addToStruct := PostProcessAction{Action: "decrypt", Field: field, Value: loadedValue}
```
Comment (KavishShah09) [inline gateway/modules/auth/match.go]:
```suggestion
			addToStruct := PostProcessAction{Action: "decrypt", Field: field}
```

File: gateway/modules/auth/match_test.go [hunk 5]
```diff
@@ -237,3 +248,159 @@ func TestMatchRemove_Rule(t *testing.T) {
...
+			wantErr: true,
+		},
+		{
+			name: "valid res",
+			m:    &Module{secret: "Olw6AhA/GzSxfhwKLxO7JJsUL6VUwwGEFTgxzoZPy9g="},
+			args: args{rule: &config.Rule{Rule: "encrypt", Fields: []string{"res.username"}}, args: map[string]interface{}{"res": map[string]interface{}{"username": "username1"}}},
+			want: &PostProcess{[]PostProcessAction{PostProcessAction{Action: "encrypt", Field: "res.username", Value: "username1"}}},
+		},
+		{
+			name:    "invalid prefix",
```
Comment (KavishShah09) [inline gateway/modules/auth/match_test.go]:
```suggestion
			name:    "invalid field prefix",
```

File: gateway/modules/auth/match.go [hunk 6]
```diff
@@ -165,3 +176,85 @@ func matchRemove(rule *config.Rule, args map[string]interface{}) (*PostProcess,
...
+	actions := &PostProcess{}
+	for _, field := range rule.Fields {
+		loadedValue, err := utils.LoadValue(field, args)
+		if err != nil {
+			logrus.Errorln("error loading value in matchEncrypt: ", err)
+			return nil, err
+		}
+		key, err := base64.StdEncoding.DecodeString(m.secret)
+		if strings.HasPrefix(field, "res") {
+			addToStruct := PostProcessAction{Action: "encrypt", Field: field, Value: loadedValue}
```
Comment (KavishShah09) [inline gateway/modules/auth/match.go]:
```suggestion
			addToStruct := PostProcessAction{Action: "encrypt", Field: field}
```

File: gateway/modules/auth/match_test.go [hunk 7]
```diff
@@ -237,3 +248,159 @@ func TestMatchRemove_Rule(t *testing.T) {
...
+		src []byte
+		key []byte
+		iv  []byte
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
```
Comment (KavishShah09) [inline gateway/modules/auth/match_test.go]: Test if encryption is actually happening

File: gateway/modules/auth/match_test.go [hunk 8]
```diff
@@ -237,3 +248,159 @@ func TestMatchRemove_Rule(t *testing.T) {
...
+		dst []byte
+		src []byte
+		key []byte
+		iv  []byte
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
```
Comment (KavishShah09) [inline gateway/modules/auth/match_test.go]: Check if decryption is actually happening

File: gateway/modules/auth/auth.go [hunk 9]
```diff
@@ -64,6 +66,7 @@ func (m *Module) SetConfig(project string, secret string, rules config.Crud, fil
 	m.project = project
 	m.rules = rules
 	m.secret = secret
+	m.aesKey, _ = base64.StdEncoding.DecodeString(aesKey)
```
Comment (YourTechBud) [inline gateway/modules/auth/auth.go]: Check for the error here

File: gateway/modules/auth/auth.go [hunk 10]
```diff
@@ -53,7 +55,7 @@ func Init(nodeID string, crud *crud.Module, schema *schema.Schema, removeProject
 }
 
 // SetConfig set the rules and secret key required by the auth block
-func (m *Module) SetConfig(project string, secret string, rules config.Crud, fileStore *config.FileStore, functions *config.ServicesModule, eventing *config.Eventing) error {
+func (m *Module) SetConfig(project string, secret string, rules config.Crud, fileStore *config.FileStore, functions *config.ServicesModule, eventing *config.Eventing, aesKey string) error {
```
Comment (YourTechBud) [inline gateway/modules/auth/auth.go]: Take the `aesKey` right after `secret`

File: gateway/modules/auth/post_action.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+					return err
+				}
+
+			case "encrypt":
+				loadedValue, err := utils.LoadValue(field.Field, map[string]interface{}{"res": doc})
+				if err != nil {
+					logrus.Errorln("error loading value in matchEncrypt: ", err)
+					return err
+				}
+				encrypted := make([]byte, len(loadedValue.(string)))
```
Comment (YourTechBud) [inline gateway/modules/auth/post_action.go]: Check if loaded value is string

File: gateway/modules/auth/post_action.go [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+					return er
+				}
+
+			case "decrypt":
+				loadedValue, err := utils.LoadValue(field.Field, map[string]interface{}{"res": doc})
+				if err != nil {
+					logrus.Errorln("error loading value in matchEncrypt: ", err)
+					return err
+				}
+				decrypted := make([]byte, len(loadedValue.(string)))
```
Comment (YourTechBud) [inline gateway/modules/auth/post_action.go]: Check if the `loadedValue` is actually string

File: gateway/modules/auth/post_action_test.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+		{testName: "Unsuccessful Test Case-invalid result", IsErrExpected: true,
+			result:      1234,
+			finalResult: 1234,
+			postProcess: &PostProcess{[]PostProcessAction{PostProcessAction{Action: "forced", Field: "res.age", Value: "1234"}}},
+		},
+		{testName: "Unsuccessful Test Case-slice of interface as result", IsErrExpected: true,
+			result:      []interface{}{1234, "suyash"},
+			finalResult: []interface{}{1234, "suyash"},
+			postProcess: &PostProcess{[]PostProcessAction{PostProcessAction{Action: "forced", Field: "res.age", Value: "1234"}}},
+		},
```
Comment (YourTechBud) [inline gateway/modules/auth/post_action_test.go]: Add your test cases here

File: gateway/modules/auth/match.go [hunk 14]
```diff
@@ -165,3 +176,83 @@ func matchRemove(rule *config.Rule, args map[string]interface{}) (*PostProcess,
...
+		if strings.HasPrefix(field, "res") {
+			addToStruct := PostProcessAction{Action: "encrypt", Field: field}
+			actions.postProcessAction = append(actions.postProcessAction, addToStruct)
+		} else if strings.HasPrefix(field, "args") {
+			loadedValue, err := utils.LoadValue(field, args)
+			if err != nil {
+				logrus.Errorln("error loading value in matchEncrypt: ", err)
+				return nil, err
+			}
+			encrypted := make([]byte, len(loadedValue.(string)))
```
Comment (YourTechBud) [inline gateway/modules/auth/match.go]: Check if the loaded value is a string

File: gateway/modules/auth/match.go [hunk 15]
```diff
@@ -165,3 +176,83 @@ func matchRemove(rule *config.Rule, args map[string]interface{}) (*PostProcess,
...
+		if strings.HasPrefix(field, "res") {
+			addToStruct := PostProcessAction{Action: "decrypt", Field: field}
+			actions.postProcessAction = append(actions.postProcessAction, addToStruct)
+		} else if strings.HasPrefix(field, "args") {
+			loadedValue, err := utils.LoadValue(field, args)
+			if err != nil {
+				logrus.Errorln("error loading value in matchEncrypt: ", err)
+				return nil, err
+			}
+			decrypted := make([]byte, len(loadedValue.(string)))
```
Comment (YourTechBud) [inline gateway/modules/auth/match.go]: Check if loaded value is a string

File: gateway/modules/auth/match_test.go [hunk 16]
```diff
@@ -237,3 +249,179 @@ func TestMatchRemove_Rule(t *testing.T) {
...
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("Module.matchDecrypt() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_encryptAESCFB(t *testing.T) {
```
Comment (YourTechBud) [inline gateway/modules/auth/match_test.go]: This test seems wrong

File: gateway/modules/auth/match_test.go [hunk 17]
```diff
@@ -237,3 +249,179 @@ func TestMatchRemove_Rule(t *testing.T) {
...
+				t.Errorf("encryptAESCFB() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if !tt.wantErr && reflect.DeepEqual(tt.args.dst, tt.args.src) {
+				t.Errorf("encryptAESCFB() encryption did not take place")
+			}
+		})
+	}
+}
+
+func Test_decryptAESCFB(t *testing.T) {
```
Comment (YourTechBud) [inline gateway/modules/auth/match_test.go]: This test seems to be wrong

**Pull request merged**